### PR TITLE
*: Use jsonnet to sync from kube-prometheus

### PR DESCRIPTION
**Pull request by brancz**
This PR starts making use of jsonnet to synchronize everything with the "upstream" kube-prometheus project. This removes a huge part of our technical debt, as for the past 1.5 years we've been synchronizing these by hand. Now we can do alerting rules, dashboards, etc. fixes once and propagate it through the entire stack by running one command. This essentially makes kube-prometheus the upstream project of the manifests in `assets/`, and overrides are defined in jsonnet code in `jsonnet/`.

This is going to take some careful testing. I already tested this on a GCE cluster and everything works wonderfully. In theory nothing should have changed, but it's hard to tell as jsonnet generates some fields in a different order which is why the diff is not perfectly readable (it does generate consistently though so in the future this will not be the case anymore).

My suggestion would be that if everyone agrees with these changes, then we merge and treat follow ups as bug fixes/regressions (if there should be any, I tried to make sure to incorporate all fixes we've done over the past few days, but it's perfectly possible that I missed something).

In the future the stack can be updated with these two commands:

```
jb install            # installs all new versions of jsonnet package dependencies
hack/build-jsonnet.sh # re-generates all of `assets/`
```

@mxinden @ironcladlou @elad661

### CODE REVIEW

**Comment by mxinden** [comment]
> ```
> jb install            # installs all new versions of jsonnet package dependencies
> hack/build-jsonnet.sh # re-generates all of `assets/`
> ```

Worked perfectly!

Would you mind making these part of `.travis.yml`?

**Comment by brancz** [comment]
> Would you mind making these part of .travis.yml?

I want to hold off on that until we have the lockfile stuff fixed, because otherwise when dependencies change we always install latest which causes frustration when building on CI as unrelated things fail (we already have this in the prometheus operator, I'd prefer to keep the maintenance burden down so we can actually focus on fixing the lockfiles).

**Comment by mxinden** [comment]
Understood. Alternative would be to not git ignore the `vendor` dir and only run `hack/build-jsonnet.sh` on Travis.

**Comment by mxinden** [comment]
This looks good to me. Very happy to see this happening.

/lgtm

**Comment by brancz** [comment]
> Understood. Alternative would be to not git ignore the vendor dir and only run hack/build-jsonnet.sh on Travis.

No strong opinion, just doesn't do much until we can check that the vendor directory is actually correct as well (a la dep ensure).

**Comment by mxinden** [comment]
@elad661 @ironcladlou I am sorry for my quick merge. I forgot that PRs by maintainers are auto-approved. Let me know if you want any direct follow-ups.

File: pkg/manifests/manifests.go [hunk 1]
```diff
@@ -406,16 +518,48 @@ func (f *Factory) PrometheusK8sRoleBinding() (*rbacv1beta1.RoleBinding, error) {
 	return rb, nil
 }
 
+func (f *Factory) PrometheusK8sRoleBindingConfig() (*rbacv1beta1.RoleBinding, error) {
+	rb, err := f.NewRoleBinding(MustAssetReader(PrometheusK8sRoleBindingConfig))
+	if err != nil {
+		return nil, err
+	}
+
+	rb.Namespace = f.namespace
```
[the remark this answers is not in the document]
Comment (brancz) [reply pkg/manifests/manifests.go]: the config binding needs to be in the namespace of the prometheus, so this is correct

File: pkg/manifests/manifests.go [hunk 2]
```diff
@@ -376,8 +470,26 @@ func (f *Factory) PrometheusK8sRoleBindingDefault() (*rbacv1beta1.RoleBinding, e
...
+}
+
 func (f *Factory) PrometheusK8sRoleDefault() (*rbacv1beta1.Role, error) {
-	return f.NewRole(MustAssetReader(PrometheusK8sRoleDefault))
+	r, err := f.NewRole(MustAssetReader(PrometheusK8sRoleDefault))
+	if err != nil {
+		return nil, err
+	}
+
+	r.Namespace = f.namespace
```
[the remark this answers is not in the document]
Comment (brancz) [reply pkg/manifests/manifests.go]: Opened https://github.com/openshift/cluster-monitoring-operator/pull/31 to fix this.

**Comment by brancz** [comment]
> Testing the timestamp/unstructured fixes required me to add RuleFiles to the ClusterRoles, which led me to discover what appear to be bugs which manifest when syncing roles and bindings.

I'm not sure I understand this. If you're not deploying the Prometheus Operator from master, then why are `RuleFile`s relevant?

**Comment by ironcladlou** [comment]
>I'm not sure I understand this. If you're not deploying the Prometheus Operator from master, then why are RuleFiles relevant?

The timestamp fix is implemented in prometheus-operator, so I was in fact deploying from master (with  my patch applied)

**Comment by brancz** [comment]
:man_facepalming: right, makes sense. Thanks for re-iterating.

**Pull request merged**